server/src/microservices/processors/metadata-extraction.processor.ts

```diff
@@ -178,165 +110,110 @@ export class MetadataExtractionProcessor {
+      id: asset.id,
+      duration: Duration.fromObject({ seconds: tags.Duration }).toFormat('hh:mm:ss.SSS'),
+      fileCreatedAt: exifData.dateTimeOriginal ?? undefined,
+    });
 
     return true;
   }
 
-  private async applyReverseGeocoding(asset: AssetEntity, newExif: ExifEntity) {
-    const { latitude, longitude } = newExif;
-    if (this.reverseGeocodingEnabled && longitude && latitude) {
-      try {
-        const { country, state, city } = await this.geocodingRepository.reverseGeocode({ latitude, longitude });
-        newExif.country = country;
-        newExif.state = state;
-        newExif.city = city;
-      } catch (error: any) {
-        this.logger.warn(
-          `Unable to run reverse geocoding due to ${error} for asset ${asset.id} at ${asset.originalPath}`,
-          error?.stack,
-        );
-      }
+  private async applyReverseGeocoding(asset: AssetEntity, exifData: ExifEntity): Promise<ExifEntity> {
+    const { latitude, longitude } = exifData;
+    if (!this.reverseGeocodingEnabled || !longitude || !latitude) {
+      return exifData;
+    }
+
+    try {
+      const geocode = await this.geocodingRepository.reverseGeocode({ latitude, longitude });
+      return { ...exifData, ...geocode };
+    } catch (error: any) {
+      this.logger.warn(
+        `Unable to run reverse geocoding due to ${error} for asset ${asset.id} at ${asset.originalPath}`,
+        error?.stack,
+      );
     }
+    return exifData;
   }
 
-  private extractDuration(duration: number | string | null) {
-    const videoDurationInSecond = Number(duration);
-    if (!videoDurationInSecond) {
+  private async exifData(asset: AssetEntity): Promise<[ExifEntity, ImmichTags]> {
+    const mediaTags = await exiftool.read<ImmichTags>(asset.originalPath).catch((error: any) => {
+      this.logger.warn(`error reading exif data (${asset.id} at ${asset.originalPath}): ${error}`, error?.stack);
       return null;
-    }
+    });
 
-    return Duration.fromObject({ seconds: videoDurationInSecond }).toFormat('hh:mm:ss.SSS');
+    const sidecarTags = asset.sidecarPath
+      ? await exiftool.read<ImmichTags>(asset.sidecarPath).catch((error: any) => {
+        this.logger.warn(`error reading exif data (${asset.id} at ${asset.sidecarPath}): ${error}`, error?.stack);
+        return null;
+      })
+      : null;
+
+    const stats = fs.statSync(asset.originalPath);
+
+    const tags = { ...mediaTags, ...sidecarTags };
+
+    const exifDate = (dt: ExifDateTime | string | undefined) => (dt instanceof ExifDateTime ? dt?.toDate() : null);
+    const zone = (dt: ExifDateTime | string | undefined) => (dt instanceof ExifDateTime ? dt.zone : null);
+
+    const parseFocalLength = (focalLength: string | undefined) =>
+      focalLength ? parseInt(focalLength.split(' ')[0]) : null;
+
+    const validate = <T>(value: T): T | null => typeof value === 'string' ? null : value ?? null;
+
+    return [
+      <ExifEntity>{
+        // altitude: tags.GPSAltitude ?? null,
```

Comment:
question still remains: altitude?

Reply:
I assume we could add it. We would need a new column for it and then decide how/when to show the data in the UI?

Reply:
Let's make this a separate PR after this has been merged.